### DISCUSSION TITLE
to_crs is slow

### DISCUSSION THREAD

**Discussion by dlasusa** [geopandas/geopandas#2870] [Q&A] [answered]
Hello all,

I'm reading in a .shp file

shape info: (20123, 6):

```py
gdf = gpd.read_file(file)
```
When I check the CRS with `gdf.crs`  it correctly states `EPSG:2230` among the rest of the output.

I then try to reproject trying each of:
```py
gdf.to_crs(4326)
gdf.to_crs(epsg=4326)
gdf['geometry'].to_crs(epsg=4326) 
```
All of these run successfully, but each takes a bit over 2mins.

When using QGIS to save the same Shapefile with reprojected coordinates, it takes less than 10secs.

Using arcpy, it takes between 10 and 20secs.

I'm wondering what I might be doing incorrectly to cause this to take over 2mins?

My environment:
Win 10 (64bit)
Jupyter Notebook inside VS Code
* I don't use conda, everything is installed via `pip`

`gpd.show_versions()` Output:
```
SYSTEM INFO
-----------
python     : 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)]
executable : [d:\Python_Projects\2023\2023-03-31_GIS_Metadata_Tools\venv\Scripts\python.exe](file:///D:/Python_Projects/2023/2023-03-31_GIS_Metadata_Tools/venv/Scripts/python.exe)
machine    : Windows-10-10.0.19045-SP0

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.11.1
GEOS lib   : None
GDAL       : 3.5.2
GDAL data dir: [d:\Python_Projects\2023\2023-03-31_GIS_Metadata_Tools\venv\Lib\site-packages\fiona\gdal_data](file:///D:/Python_Projects/2023/2023-03-31_GIS_Metadata_Tools/venv/Lib/site-packages/fiona/gdal_data)
PROJ       : 9.2.0
PROJ data dir: [d:\Python_Projects\2023\2023-03-31_GIS_Metadata_Tools\venv\Lib\site-packages\pyproj\proj_dir\share\proj](file:///D:/Python_Projects/2023/2023-03-31_GIS_Metadata_Tools/venv/Lib/site-packages/pyproj/proj_dir/share/proj)

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.12.2
numpy      : 1.24.2
pandas     : 1.5.3
pyproj     : 3.5.0
shapely    : 2.0.1
fiona      : 1.9.2
geoalchemy2: None
geopy      : None
matplotlib : None
mapclassify: None
pygeos     : None
pyogrio    : 0.5.1
psycopg2   : None
pyarrow    : None
rtree      : 1.0.1
```

Thanks for any help or insight!

**Comment by jdmcbr** [comment]
@dlasusa Can you share the shapefile where you're seeing this performance?

**Comment by dlasusa** [comment]
@jdmcbr I wasn't sure if you needed JUST the shapefile or the rest as well, so I zipped them up and included all of them.  Thanks!
[SD_County_Bike_Routes_2022_04.zip](https://github.com/geopandas/geopandas/files/11213533/SD_County_Bike_Routes_2022_04.zip)

**Reply by martinfleis** [reply]
There's something strange happening in your environment. On my machine with the env equal to yours, the `to_crs` call takes about 330 ms, in QGIS the same takes about 2 secs. But I don't have a windows machine to test on.

I am not entirely sure how to debug this. Can you try creating a new fresh environment to see if the issue persists? It could be some installation issue.

**Reply by theroggy** [reply]
I also had a very quick look because I'm working on windows... and it takes ~0.5 seconds per to_crs` call... 

I used an existing env, so not 100% same, but... at least it doesn't seem to be a general windows issue...

**Reply by dlasusa** [reply]
Thanks to you both.  Are either or both of you using pip? for your env/install?  I just did a new env with the following steps:

```
Create venv (python 3.11.2)
update pip
pip install fiona
pip install pyproj
pip install rtree
pip install shapely
pip install geopandas
```
Then in my notebook:

```py
import geopandas as gpd
bike_gdf = gpd.read_file(path_to_file)
bike_gdf.crs # just to confirm it was read ok
bike_gdf.to_crs(4326)
```
Took 1min 58secs

I don't really use conda, but might try miniconda just to try something different.  But if either of you are using pip, it makes me think this should work via pip.

Thanks!

**Reply by dlasusa** [reply]
Maybe some progress?

I duplicated my steps on a linux laptop I have (though it's running Python 3.10.6)

and `to_crs` took 48secs.

50% speed up! 

Though that's still much slower than others are seeing.

**Reply by theroggy** [reply]
I'm using conda, or more specific [mambaforge](https://github.com/conda-forge/miniforge#mambaforge) because miniconda isn't free anymore for non-private use since some time...

**Reply by dlasusa** [reply]
More testing results:

Using miniconda when I do a `df.crs` I get (which looks a bit different vs. my normal pip env - pip version at bottom for reference):

```
<Projected CRS: PROJCS["NAD83 / California zone 6 (ftUS)",GEOGCS[" ...>
Name: NAD83 / California zone 6 (ftUS)
Axis Info [cartesian]:
- [east]: Easting (US survey foot)
- [north]: Northing (US survey foot)
Area of Use:
- undefined
Coordinate Operation:
- name: unnamed
- method: Lambert Conic Conformal (2SP)
Datum: North American Datum 1983
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich
```
Then when I try to run

```py
df.to_crs(4326)
```
I get:

```
ProjError                                 Traceback (most recent call last)
 in 
----> 1 bike_df.to_crs(4326)

~\miniconda3\envs\test_geopandas\lib\site-packages\geopandas\geodataframe.py in to_crs(self, crs, epsg, inplace)
   1362         else:
   1363             df = self.copy()
-> 1364         geom = df.geometry.to_crs(crs=crs, epsg=epsg)
   1365         df.geometry = geom
   1366         if not inplace:

~\miniconda3\envs\test_geopandas\lib\site-packages\geopandas\geoseries.py in to_crs(self, crs, epsg)
   1122         """
   1123         return GeoSeries(
-> 1124             self.values.to_crs(crs=crs, epsg=epsg), index=self.index, name=self.name
   1125         )
   1126 

~\miniconda3\envs\test_geopandas\lib\site-packages\geopandas\array.py in to_crs(self, crs, epsg)
    777         transformer = Transformer.from_crs(self.crs, crs, always_xy=True)
    778 
--> 779         new_data = vectorized.transform(self.data, transformer.transform)
    780         return GeometryArray(new_data, crs=crs)
    781 

~\miniconda3\envs\test_geopandas\lib\site-packages\geopandas\_vectorized.py in transform(data, func)
   1112 
   1113         coords_z = get_coordinates(data[has_z], include_z=True)
-> 1114         new_coords_z = func(coords_z[:, 0], coords_z[:, 1], coords_z[:, 2])
   1115         result[has_z] = set_coordinates(data[has_z].copy(), np.array(new_coords_z).T)
   1116 

~\miniconda3\envs\test_geopandas\lib\site-packages\pyproj\transformer.py in transform(self, xx, yy, zz, tt, radians, errcheck, direction)
    428             intime = None
    429         # call pj_transform.  inx,iny,inz buffers modified in place.
--> 430         self._transformer._transform(
    431             inx,
    432             iny,

pyproj/_transformer.pyx in pyproj._transformer._Transformer._transform()

ProjError: x, y, z, and time must be same size
```
PIP `.crs` info:
```
<Projected CRS: EPSG:2230>
Name: NAD83 / California zone 6 (ftUS)
Axis Info [cartesian]:
- [east]: Easting (US survey foot)
- [north]: Northing (US survey foot)
Area of Use:
- undefined
Coordinate Operation:
- name: unnamed
- method: Lambert Conic Conformal (2SP)
Datum: North American Datum 1983
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich
```

**Reply by martinfleis** [reply]
Do you have the same version of pyproj and PROJ in venv/pip and conda env?

**Reply by theroggy** [reply]
Odd. I wonder if your conda environment is properly activated. Are you still running within a notebook? I'm running it directly as a script (in VSCode)?

I put the script I ran to test [here](https://github.com/theroggy/pysnippets/blob/main/pysnippets/geopandas_questions/to_crs_slow.py).

**Reply by dlasusa** [reply]
All notebook runs so far.  In my post above I grouped the following together:

```py
import geopandas as gpd
bike_gdf = gpd.read_file(path_to_file)
bike_gdf.crs # just to confirm it was read ok
bike_gdf.to_crs(4326)
```
But these were actually run as separate cells in a notebook. (in VS Code).

Let me check the script you shared.  Thanks!

**Reply by dlasusa** [reply]
@martinfleis My 2 environments:

Miniconda:
```
SYSTEM INFO
-----------
python     : 3.9.16 (main, Mar  8 2023, 10:39:24) [MSC v.1916 64 bit (AMD64)]
executable : ~\miniconda3\envs\test_geopandas\python.exe
machine    : Windows-10-10.0.19045-SP0

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.8.0
GEOS lib   : None
GDAL       : 3.6.2
GDAL data dir: None
PROJ       : 6.2.1
PROJ data dir: ~\miniconda3\envs\test_geopandas\Library\share\proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.12.2
numpy      : 1.23.5
pandas     : 1.5.3
pyproj     : 2.6.1.post1
shapely    : 2.0.1
fiona      : 1.9.1
geoalchemy2: None
geopy      : None
matplotlib : 3.7.1
mapclassify: 2.5.0
pygeos     : None
pyogrio    : None
psycopg2   : None
pyarrow    : None
rtree      : 1.0.1
```

PIP:
```
SYSTEM INFO
-----------
python     : 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)]
executable : D:\Python_Projects\2023\2023-04-13_geopandas_troubleshooting\venv\Scripts\python.exe
machine    : Windows-10-10.0.19045-SP0

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.11.1
GEOS lib   : None
GDAL       : 3.5.2
GDAL data dir: D:\Python_Projects\2023\2023-04-13_geopandas_troubleshooting\venv\Lib\site-packages\fiona\gdal_data
PROJ       : 9.2.0
PROJ data dir: D:\Python_Projects\2023\2023-04-13_geopandas_troubleshooting\venv\Lib\site-packages\pyproj\proj_dir\share\proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.12.2
numpy      : 1.24.2
pandas     : 2.0.0
pyproj     : 3.5.0
shapely    : 2.0.1
fiona      : 1.9.3
geoalchemy2: None
geopy      : None
matplotlib : None
mapclassify: None
pygeos     : None
pyogrio    : None
psycopg2   : None
pyarrow    : 11.0.0
rtree      : 1.0.1
```
@theroggy When I ran your script (in my PIP env) I got the following output:
```
running gdf.to_crs(4326) took 0:02:00.857835
running gdf.to_crs(epsg=4326) took 0:02:02.569396
running gdf.geometry.to_crs(epsg=4326) took 0:02:03.123322
```
When I ran it in my miniconda env I get the same error I posted above (here is a shorter version):
```
pyproj.exceptions.ProjError: x, y, z, and time must be same size
```

**Reply by martinfleis** [reply]
How did you install pyproj to your conda env? You have an awfully old version of pyproj and PROJ there.

**Reply by dlasusa** [reply]
With conda (which I'm admittedly less familiar with for envs) I used the install page: https://geopandas.org/en/stable/getting_started/install.html

and did this: 

```py
conda install geopandas
```
I didn't actually install anything individually.

**Reply by martinfleis** [reply, marked as answer]
Can you try specifying the channel?

```
conda install geopandas -c conda-forge
```

**Reply by dlasusa** [reply]
Sure thing.  Let me do a fresh env.  I'm also trying this on a machine in the office (remote desktop).  I definitely think it's something I'm doing (or not doing) in my setup.  My times in the office are ~90secs.  So whatever I'm doing, I'm doing it consistently.

**Reply by dlasusa** [reply]
SUCCESS!

Creating a new env and using:
```py
conda install geopandas -c conda-forge
```
Which provided the following environment:
```
SYSTEM INFO
-----------
python     : 3.11.3 | packaged by conda-forge | (main, Apr  6 2023, 08:50:54) [MSC v.1934 64 bit (AMD64)]
executable : ~\miniconda3\envs\test2_crs\python.exe
machine    : Windows-10-10.0.19045-SP0

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.11.2
GEOS lib   : None
GDAL       : 3.6.3
GDAL data dir: None
PROJ       : 9.2.0
PROJ data dir: ~\miniconda3\envs\test2_crs\Library\share\proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.12.2
numpy      : 1.24.2
pandas     : 2.0.0
pyproj     : 3.5.0
shapely    : 2.0.1
fiona      : 1.9.3
geoalchemy2: None
geopy      : None
matplotlib : 3.7.1
mapclassify: 2.5.0
pygeos     : None
pyogrio    : None
psycopg2   : None
pyarrow    : None
rtree      : 1.0.1
```
Then running the script @theroggy provided I'm now getting the following times:
```
running gdf.to_crs(4326) took 0:00:01.834259
running gdf.to_crs(epsg=4326) took 0:00:00.692003
running gdf.geometry.to_crs(epsg=4326) took 0:00:00.733001
```
I'll mark the install from conda-forge as the answer, though I'm curious if we know which package (or packages?) may have been causing the issue?  I may still fight with this on my own using pip for my env, but at least that points me in the right direction. 

Thanks a bunch for all the help!

**Reply by martinfleis** [reply]
Most likely PROJ, exposed via pyproj was the issue.

**Reply by martinfleis** [reply]
@snowman2 do you have any idea why we see this behaviour? Is there something disabled or different in the wheel on PyPI? Or maybe a bug?

**Reply by snowman2** [reply]
Possibly related: https://github.com/pyproj4/pyproj/issues/1268

**Reply by dlasusa** [reply]
Not sure it's helpful, and I know a pic might be hard to read/use, but I've slowly tried to update my non-working env to slowly match the working env.  

Green title: Working
Red titles: Slow
Yellow cells: What's still different from the working env
![geopandas_testing](https://user-images.githubusercontent.com/36902797/231875708-c435970d-07e3-4cc6-8528-6f1d02259731.png)

I'm not sure how to update GDAL.  I downloaded and installed GDAL 3.6.3 - and it shows up as the right version in my add/remove programs, but the env doesn't seem to see it.

**Reply by m-richards** [reply]
Another data point, I have an existing windows setup with allpackages installed via pip, with pyproj 3.4.1 in the environment, to_crs takes 0.3 seconds

**Reply by dlasusa** [reply]
Thanks @m-richards -  In the last bit of my testing for the day, I went back to a project I was working on in March and it still has pyproj 3.4.1.

So I copied the data and the script into the folder and ran it "as is" (no package updates).  And it ran in:

```
running gdf.to_crs(4326) took 0:00:00.625900
```
So now I'm almost more curious why my one successful run from earlier (with pyproj 3.5.0 worked!?)  but for now, that'll have to be tomorrow's problem lol

I appreciates everyone's help and data points today!

**Reply by theroggy** [reply]
It sounds like a confirmation/an example of the performance regression from 3.4.1 to 3.5.0 mentioned by @snowman2:
`Possibly related:` https://github.com/pyproj4/pyproj/issues/1268

So, specifying "pyproj<3.5" when installing using pip will probably do the trick

**Reply by jorisvandenbossche** [reply]
That indeed sounds like a good candidate for explaining the performance issue (the type of projection also matches), but the strange thing is that this seemingly didn't appear with pyproj 3.5 / PROJ 9.2 with conda-forge (https://github.com/geopandas/geopandas/discussions/2870#discussioncomment-5607772)
